### PR TITLE
feat: [DEL-5731]: New API for fetching terraform example module file and store delegate feedbacks

### PR DESCRIPTION
**Pull request by xingchi-jin**
## Describe your changes
We started our terraform journey as a command line based approach which was
1. Curl a terraform main.tf file without arguments - like token, managerUrl ..
2. Terraform init
3. Terraform apply -var .. - var ...

This doesn't work because these variables are not defined in the local main.tf customer downloads. We have to provide a main.tf file that have variables populated in it

This PR provides an API that gives the main.tf file.

As part of the same PR, it provides a new API to store delegate installation feedbacks 

Test Status: 
Local Done
PR ENV Done

Test results
1. Feedback API
image

2. Improved Docker command
image

3. Improved Helm command
Screen Shot 2023-01-10 at 4 59 10 PM

4. fetch terraform file from API
Screen Shot 2023-01-10 at 5 04 10 PM

5. Deploy this terraform file
Screen Shot 2023-01-10 at 5 06 22 PM
Screen Shot 2023-01-10 at 5 11 34 PM

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/42455)

### CODE REVIEW

**Comment by abhayganvir** [comment]
trigger  test_copyrightcheck

**Comment by abhayganvir** [comment]
trigger smartchecks

**Review by raghuAtWings — APPROVED** [review]
> Reviewed 4 of 14 files at r1, 6 of 7 files at r2, 5 of 5 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42455)* status: :shipit: complete! ✓

**Comment by xingchi-jin** [comment]
trigger smartchecks

File: 420-delegate-service/src/main/java/io/harness/delegate/beans/DelegateFeedbackMapper.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * Use of this source code is governed by the PolyForm Free Trial 1.0.0 license
+ * that can be found in the licenses directory at the root of this repository, also available at
+ * https://polyformproject.org/wp-content/uploads/2020/05/PolyForm-Free-Trial-1.0.0.txt.
+ */
+
+package io.harness.delegate.beans;
+
+public class DelegateFeedbackMapper {
+  public static DelegateFeedback map(final DelegateFeedbackDTO delegateFeedbackDTO) {
+    return DelegateFeedback.builder()
```
Comment (meenakshi-harness) [inline 420-delegate-service/src/main/java/io/harness/delegate/beans/DelegateFeedbackMapper.java:12]: also add UT for this?

File: 420-delegate-service/src/main/java/io/harness/delegate/beans/DelegateFeedbackDTO.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+@Builder
+@JsonIgnoreProperties(ignoreUnknown = true)
+@OwnedBy(DEL)
+@Slf4j
+public class DelegateFeedbackDTO {
+  private String accountId;
+  private String projectId;
+  private String orgId;
+  private String delegateName;
+  private String delegateType;
```
Comment (meenakshi-harness) [inline 420-delegate-service/src/main/java/io/harness/delegate/beans/DelegateFeedbackDTO.java:29]: Add  validation annotation where applicable.

File: 420-delegate-service/src/main/java/io/harness/delegate/beans/DelegateFeedback.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+@Data
+@Builder
+@FieldNameConstants(innerTypeName = "DelegateFeedbackKeys")
+@StoreIn(DbAliases.HARNESS)
+@Entity(value = "delegateFeedbacks", noClassnameStored = true)
+@HarnessEntity(exportable = true)
+@OwnedBy(HarnessTeam.DEL)
+public class DelegateFeedback implements PersistentEntity, UuidAware, CreatedAtAware {
+    @Id @NotNull private String uuid;
+    private String accountId;
```
Comment (meenakshi-harness) [inline 420-delegate-service/src/main/java/io/harness/delegate/beans/DelegateFeedback.java:37]: Are all these fields optional ? No validation?

File: 420-delegate-service/src/main/java/io/harness/delegate/beans/DelegateFeedback.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+@Data
+@Builder
+@FieldNameConstants(innerTypeName = "DelegateFeedbackKeys")
+@StoreIn(DbAliases.HARNESS)
+@Entity(value = "delegateFeedbacks", noClassnameStored = true)
+@HarnessEntity(exportable = true)
+@OwnedBy(HarnessTeam.DEL)
+public class DelegateFeedback implements PersistentEntity, UuidAware, CreatedAtAware {
```
Comment (meenakshi-harness) [inline 420-delegate-service/src/main/java/io/harness/delegate/beans/DelegateFeedback.java:35]: Add TypeAlias to entity?

File: 400-rest/src/main/java/io/harness/delegate/service/impl/DelegateInstallationCommandServiceImpl.java [hunk 5]
```diff
@@ -64,38 +67,56 @@ public DelegateInstallationCommandServiceImpl(
...
+        .put("manager_url", managerUrl)
+        .put("image", image)
+        .build();
+  }
+
+  private String getDefaultNgToken(String accountId) {
+    final Map<String, Boolean> activeStatus =
+        delegateNgTokenService.isDelegateTokenActive(accountId, List.of(DEFAULT_TOKEN_NAME));
+    if (Objects.isNull(activeStatus) || activeStatus.isEmpty()) {
+      final String errorMsg = String.format("default token not found for account %s.", accountId);
```
Comment (meenakshi-harness) [inline 400-rest/src/main/java/io/harness/delegate/service/impl/DelegateInstallationCommandServiceImpl.java:114]: Default token..

File: 400-rest/src/main/java/io/harness/delegate/service/impl/DelegateInstallationCommandServiceImpl.java [hunk 6]
```diff
@@ -64,38 +67,56 @@ public DelegateInstallationCommandServiceImpl(
...
+    return substitute.replace(content);
+  }
+
+  private Map<String, String> getScriptParams(
+      final String managerUrl, final String accountId, final String tokenValue, final String image) {
+    return ImmutableMap.<String, String>builder()
+        .put("account_id", accountId)
+        .put("token", tokenValue)
+        .put("manager_url", managerUrl)
+        .put("image", image)
```
Comment (meenakshi-harness) [inline 400-rest/src/main/java/io/harness/delegate/service/impl/DelegateInstallationCommandServiceImpl.java:106]: Move all these strings to constants.

File: 400-rest/src/main/java/io/harness/delegate/service/impl/DelegateInstallationCommandServiceImpl.java [hunk 7]
```diff
@@ -64,38 +67,56 @@ public DelegateInstallationCommandServiceImpl(
...
   @Override
   public String getCommand(@NotBlank String commandType, @NotBlank String managerUrl, @NotBlank String accountId) {
-    final Map<String, Boolean> activeStatus =
-        delegateNgTokenService.isDelegateTokenActive(accountId, List.of(DEFAULT_TOKEN_NAME));
-    if (Objects.isNull(activeStatus) || activeStatus.isEmpty()) {
-      final String errorMsg = String.format("default token not found for account %s.", accountId);
-      log.error(errorMsg);
-      throw new DelegateTokenNotFoundException(errorMsg);
-    }
+    final String tokenValue = getDefaultNgToken(accountId);
```
Comment (meenakshi-harness) [inline 400-rest/src/main/java/io/harness/delegate/service/impl/DelegateInstallationCommandServiceImpl.java:70]: UT for this?

File: 420-delegate-service/src/main/java/io/harness/delegate/beans/DelegateFeedbackDTO.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+@Getter
+@Builder
+@JsonIgnoreProperties(ignoreUnknown = true)
+@OwnedBy(DEL)
+@Slf4j
+public class DelegateFeedbackDTO {
+  private String accountId;
+  private String projectId;
+  private String orgId;
```
Comment (meenakshi-harness) [inline 420-delegate-service/src/main/java/io/harness/delegate/beans/DelegateFeedbackDTO.java:27]: We follow accountIdentifier, orgIdentifier and projectIdentifier for other NG DTOs and Entity.

File: 420-delegate-service/src/main/java/io/harness/delegate/beans/DelegateFeedback.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    @Id @NotNull private String uuid;
+    private String accountId;
+    private String projectId;
+    private String orgId;
+    private String delegateName;
+    private String delegateType;
```
Comment (meenakshi-harness) [inline 420-delegate-service/src/main/java/io/harness/delegate/beans/DelegateFeedback.java:41]: Why delegate type is not Enum but String?

**Comment by xingchi-jin** [comment]
As discussed, I will address the comments in another PR https://harness.atlassian.net/browse/DEL-5802

**Review by meenakshi-harness — APPROVED** [review]
Approving on request.

> Reviewed 3 of 14 files at r1, 4 of 7 files at r2, 4 of 5 files at r3, 1 of 1 files at r4, 3 of 3 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42455#-:-NLxx0t37wW-vL1HI71Y:bsawx8l)* status: :shipit: complete! ✓

**Review by raghuAtWings — APPROVED** [review]
> Reviewed 3 of 3 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42455)* status: :shipit: complete! ✓

**Review by Vbhatia20 — APPROVED** [review]
Please ensure to validate this.

**Pull request merged**